crates/oxc_linter/src/rules/jest/prefer_hooks_in_order.rs

```diff
@@ -0,0 +1,1271 @@
+    ///     afterAll(() => {
+    ///         removeMyDatabase();
+    ///     });
+    /// });
+    /// ```
+    ///
+    /// ```javascript
+    /// // valid
+    /// describe('foo', () => {
+    ///     beforeAll(() => {
+    ///         createMyDatabase();
+    ///     });
+    ///
+    ///     beforeEach(() => {
+    ///         seedMyDatabase();
+    ///     });
+    ///
+    ///     it('accepts this input', () => {
+    ///         // ...
+    ///     });
+    ///     it('returns that value', () => {
+    ///         // ...
+    ///     });
+    ///     describe('when the database has specific values', () => {
+    ///         const specificValue = '...';
+    ///         beforeEach(() => {
+    ///             seedMyDatabase(specificValue);
+    ///         });
+    ///         it('accepts that input', () => {
+    ///             // ...
+    ///         });
+    ///         it('throws an error', () => {
+    ///             // ...
+    ///         });
+    ///         beforeEach(() => {
+    ///             mockLogger();
+    ///         });
+    ///         afterEach(() => {
+    ///             clearLogger();
+    ///         });
+    ///         it('logs a message', () => {
+    ///             // ...
+    ///         });
+    ///     });
+    ///     afterAll(() => {
+    ///         removeMyDatabase();
+    ///     });
+    /// });
+    /// ```
+    ///
+    ///
+    /// This rule is compatible with [eslint-plugin-vitest](https://github.com/veritem/eslint-plugin-vitest/blob/main/docs/rules/prefer-hooks-in-order.md),
+    /// to use it, add the following configuration to your `.eslintrc.json`:
+    ///
+    /// ```json
+    /// {
+    ///   "rules": {
+    ///      "vitest/prefer-hooks-in-order": "error"
+    ///   }
+    /// }
+    PreferHooksInOrder,
+    style,
+);
+
+impl Rule for PreferHooksInOrder {
+    fn run_once(&self, ctx: &LintContext) {
+        let mut hook_groups: FxHashMap<ScopeId, Vec<AstNode>> = FxHashMap::default();
+
+        for node in ctx.nodes().iter() {
+            hook_groups.entry(node.scope_id()).or_default().push(*node);
+        }
+
+        for (_, nodes) in hook_groups {
+            let mut previous_hook_index = -1;
+
+            for node in nodes {
+                if let AstKind::CallExpression(call_expr) = node.kind() {
+                    let possible_jest_node = &PossibleJestNode { node: &node, original: None };
+                    Self::check(&mut previous_hook_index, possible_jest_node, call_expr, ctx);
+                };
+            }
+        }
+    }
+}
+
+impl PreferHooksInOrder {
+    fn check<'a>(
+        previous_hook_index: &mut i32,
+        possible_jest_node: &PossibleJestNode<'a, '_>,
+        call_expr: &'a CallExpression<'_>,
+        ctx: &LintContext<'a>,
+    ) {
+        let Some(ParsedJestFnCallNew::GeneralJestFnCall(jest_fn_call)) =
+            parse_jest_fn_call(call_expr, possible_jest_node, ctx)
+        else {
+            *previous_hook_index = -1;
```

Comment:
Curious about why we `-1` here.

Reply:
Because of these two cases:

```js
describe('my test', () => {
    afterEach(() => {});
    afterAll(() => {});

    doSomething();

    beforeAll(() => {});
    beforeEach(() => {});
});

describe('my test', () => {
    afterEach(() => {});
    afterAll(() => {});

    it('test', () => {});

    beforeAll(() => {});
    beforeEach(() => {});
  });
```

After a non-hook call or non-jest call, it need to reset the index, or it would fail.